src/gss_names.c

```diff
@@ -281,6 +281,80 @@ uint32_t gssntlm_import_name(uint32_t *minor_status,
                                        output_name);
 }
 
+size_t gssntlm_get_attrs_count(const struct gssntlm_name_attribute *attrs)
+{
+    size_t attrs_count = 0;
+    if (!attrs) return 0;
+    while (attrs[attrs_count].attr_name != NULL) attrs_count++;
+
+    return attrs_count;
+}
+
+int gssntlm_copy_attrs(const struct gssntlm_name_attribute *src,
+                       struct gssntlm_name_attribute **dst)
+{
+    struct gssntlm_name_attribute *copied_attrs = NULL;
+
+    size_t attrs_count = gssntlm_get_attrs_count(src);
+    if (attrs_count == 0) {
+        *dst = NULL;
+        return 0;
+    }
+
+    copied_attrs = malloc(sizeof(struct gssntlm_name_attribute) *
+                          (attrs_count + 1)); /* +1 for terminator entry */
+    if (copied_attrs == NULL) {
+        return ENOMEM;
+    }
+
+    for (size_t i = 0; i < attrs_count; i++) {
+        /* read-only persistent data ptr can be just copied */
+        copied_attrs[i].attr_name = src[i].attr_name;
+        copied_attrs[i].attr_value.length = src[i].attr_value.length;
+        copied_attrs[i].attr_value.value = malloc(src[i].attr_value.length);
+        if (copied_attrs[i].attr_value.value == NULL) {
+            /* release all previously allocated memory */
+            while (i > 0) {
+                free(copied_attrs[--i].attr_value.value);
+            }
+            free(copied_attrs);
+            return ENOMEM;
+        }
+        memcpy(copied_attrs[i].attr_value.value, src[i].attr_value.value,
+               src[i].attr_value.length);
+    }
+
+    *dst = copied_attrs;
+    return 0;
+}
+
+/* Returns index of attribute in the array or -1 if not found */
+int gssntlm_find_attr(const struct gssntlm_name_attribute *attrs,
+                      const char *attr_name, size_t attr_name_len)
+{
+    int i = 0;
+    if (!attrs) return -1;
+    while (attrs[i].attr_name != NULL) {
+        if (strncasecmp(attrs[i].attr_name, attr_name, attr_name_len) == 0) {
+            return i;
+        }
+        i++;
+    }
+    return -1;
+}
+
+void gssntlm_release_attrs(struct gssntlm_name_attribute **attrs)
+{
+    size_t attrs_count = gssntlm_get_attrs_count(*attrs);
```

Comment:
why use the counter when you can simply iterate over attrs until you find a zero attribute?

Reply:
Because gssntlm_get_attrs_count handles the case when *attrs is zero (no attrs to iterate at all).
Ok, if you don't like this, let's do this check in-place...